frontend/src/routes/BillabilityMonthlyWorkingHours.jsx

```diff
@@ -0,0 +1,166 @@
+    setSelectedYear(e.target.value);
+  };
+
+  const applyFilters = () => {
+    axios
+      .get(
+        `http://localhost:3000/billability-working-hours/${selectedYear}/${selectedYearsBack}`
+      )
+      .then((response) => {
+        const transformedTableData = {};
+        const rawTableData = response.data.data;
+
+        for (let i = 0; i < rawTableData.length; i++) {
+          const year = rawTableData[i].year;
+          const month = rawTableData[i].month;
+          const total_hours = rawTableData[i].total_hours;
+          const billable = rawTableData[i].billable;
+
+          if (!transformedTableData[year]) {
+            transformedTableData[year] = { year: year, total: 0 };
+          }
+          if (billable === 1) {
+            transformedTableData[year][month] =
+              (transformedTableData[year][month] || 0) + total_hours;
+            transformedTableData[year].total += total_hours;
+          }
+        }
+
+        setBillableTableData(Object.values(transformedTableData));
+
+        const transformedNonBillableTableData = {};
+        const rawNonBillableTableData = response.data.data;
+
+        for (let i = 0; i < rawNonBillableTableData.length; i++) {
+          const year = rawNonBillableTableData[i].year;
+          const month = rawNonBillableTableData[i].month;
+          const total_hours = rawNonBillableTableData[i].total_hours;
+          const billable = rawNonBillableTableData[i].billable;
+
+          if (!transformedNonBillableTableData[year]) {
+            transformedNonBillableTableData[year] = { year: year, total: 0 };
+          }
+          if (billable === 0) {
+            transformedNonBillableTableData[year][month] =
+              (transformedNonBillableTableData[year][month] || 0) + total_hours;
+            transformedNonBillableTableData[year].total += total_hours;
+          }
+        }
+
+        setNonBillableTableData(Object.values(transformedNonBillableTableData));
+
+        const rawData = response.data.data;
+        const transformedData = {};
+
+        for (let j = 0; j < rawData.length; j++) {
+          const graphYear = rawData[j].year;
+          const graphMonth = rawData[j].month;
+          const graphTotal_hours = rawData[j].total_hours;
+          const graphBillable = rawData[j].billable;
+          if (!transformedData[graphMonth]) {
+            transformedData[graphMonth] = { month: graphMonth };
+          }
+          if (!transformedData[graphMonth][graphYear]) {
+            transformedData[graphMonth][graphYear] = {
+              billable: 0,
+              non_billable: 0,
+            };
+          }
+          if (graphBillable === 1) {
+            transformedData[graphMonth][graphYear].billable = graphTotal_hours;
+          } else {
+            transformedData[graphMonth][graphYear].non_billable =
+              graphTotal_hours;
+          }
+        }
+        setGraphData(Object.values(transformedData));
```

Comment:
This part is quite repetitive. Can you think of how you could use a single for loop (instead of three) to transform data for each use case? What variables are actually the same?

Reply:
Great improvements!